JitArm64: Never check downcount on block entry

JosJuice opened the pull request:
Jumping between linked blocks currently works as follows: First, at the end of the first block, we check if the downcount is greater than zero. If it is, we jump to the `normalEntry` of the block. So far so good. But if the downcount wasn't greater than zero, we jump to the `checkedEntry` of the block, which checks the downcount *again* and then jumps to `do_timing` if it's less than zero (which seems like an off by one error – Jit64 doesn't do anything like this). This second check is rather redundant. Let's jump to `do_timing` where we previously jumped to `checkedEntry`.

Jit64 doesn't check the downcount on block entry. See PR #7634.

dvessel commented:
Console error & backtrace when starting Rogue Leader. Error happens at startup.

```
54:20:151 Core/PowerPC/JitArm64/Jit.cpp:440 E[MASTER]: Warning: An error occurred.

  Condition: GetCodePtr() == farcode_2_addr
  File: /Users/foo/dolphin/Source/Core/Core/PowerPC/JitArm64/Jit.cpp
  Line: 440
  Function: WriteExit

Ignore and continue?
bt
Process 47151 stopped
* thread #30, name = 'CPU-GPU thread', stop reason = EXC_BREAKPOINT (code=1, subcode=0x100baa868)
    frame #0: 0x0000000100baa868 Dolphin`JitArm64::WriteExit(unsigned int, bool, unsigned int) + 776
Dolphin`JitArm64::WriteExit:
->  0x100baa868 <+776>: brk    #0x1
    0x100baa86c <+780>: b      0x100baa870               ; <+784>
    0x100baa870 <+784>: b      0x100baa874               ; <+788>
    0x100baa874 <+788>: add    x0, x19, #0x1b0
Target 0: (Dolphin) stopped.
(lldb) bt
* thread #30, name = 'CPU-GPU thread', stop reason = EXC_BREAKPOINT (code=1, subcode=0x100baa868)
  * frame #0: 0x0000000100baa868 Dolphin`JitArm64::WriteExit(unsigned int, bool, unsigned int) + 776
    frame #1: 0x0000000100bc240c Dolphin`JitArm64::bcx(UGeckoInstruction) + 652
    frame #2: 0x0000000100bf2b14 Dolphin`JitArm64::CompileInstruction(PPCAnalyst::CodeOp&) + 104
    frame #3: 0x0000000100ba5ac0 Dolphin`JitArm64::DoJit(unsigned int, JitBlock*, unsigned int) + 2596
    frame #4: 0x0000000100ba4750 Dolphin`JitArm64::Jit(unsigned int, bool) + 764
    frame #5: 0x0000000100ba3ee4 Dolphin`JitArm64::Jit(unsigned int) + 40
    frame #6: 0x0000000100aece64 Dolphin`JitTrampoline(JitBase&, unsigned int) + 40
    frame #7: 0x0000000170d510b4
```

JosJuice commented:
Thanks! Fixed.

dvessel commented:
I brought this up on discord. Posting it here so it doesn’t get lost.

RS2 freezes when switching to the targeting computer _(Y Button)_. Usually doesn’t freeze but this is from a dtm file that can catch it every time. I can share it if needed.

https://youtu.be/Oq__l07W93I?t=17

[And the stack trace](https://gist.github.com/dvessel/35a19464e4d469fd43683dd2ef5b7053)

JosJuice commented:
@dvessel Can you check if this patch fixes it?

```
diff --git a/Source/Core/Core/PowerPC/JitArm64/JitArm64Cache.cpp b/Source/Core/Core/PowerPC/JitArm64/JitArm64Cache.cpp
index 512f211037..18a6f34c39 100644
--- a/Source/Core/Core/PowerPC/JitArm64/JitArm64Cache.cpp
+++ b/Source/Core/Core/PowerPC/JitArm64/JitArm64Cache.cpp
@@ -71,7 +71,11 @@ void JitArm64BlockCache::WriteLinkBlock(Arm64Gen::ARM64XEmitter& emit,
   // Use a fixed number of instructions so we have enough room for any patching needed later.
   const u8* end = start + BLOCK_LINK_SIZE;
   while (emit.GetCodePtr() < end)
+  {
     emit.BRK(101);
+    if (emit.HasWriteFailed())
+      return;
+  }
   ASSERT(emit.GetCodePtr() == end);
 }
```

The reason why I didn't push this patch to the pull request is because if I were to push without rebasing, the buildbot would get upset, and if I were to rebase, your DTM might not sync anymore. If you can confirm that this fixes it, I'll rebase and push.

dvessel commented:
Rebased the PR and still freezes. Applied patch and now getting this error at the same point where it freezes:

The dtm file doesn’t depend on a save state so rebasing is fine.

```
12:01:858 Core/PowerPC/JitArm64/Jit.cpp:476 E[MASTER]: Warning: An error occurred.

  Condition: GetCodePtr() == farcode_1_addr
  File: dolphin/Source/Core/Core/PowerPC/JitArm64/Jit.cpp
  Line: 476
  Function: WriteExit

Ignore and continue?

Process 29811 stopped
* thread #33, name = 'CPU-GPU thread', stop reason = EXC_BREAKPOINT (code=1, subcode=0x100be42ec)
    frame #0: 0x0000000100be42ec Dolphin`JitArm64::WriteExit(unsigned int, bool, unsigned int) + 544
Dolphin`JitArm64::WriteExit:
->  0x100be42ec <+544>: brk    #0x1
    0x100be42f0 <+548>: b      0x100be42f4               ; <+552>
    0x100be42f4 <+552>: b      0x100be42f8               ; <+556>
    0x100be42f8 <+556>: add    x0, x19, #0x1b0
Target 0: (Dolphin) stopped.

(lldb) bt
* thread #33, name = 'CPU-GPU thread', stop reason = EXC_BREAKPOINT (code=1, subcode=0x100be42ec)
  * frame #0: 0x0000000100be42ec Dolphin`JitArm64::WriteExit(unsigned int, bool, unsigned int) + 544
    frame #1: 0x0000000100bfcb14 Dolphin`JitArm64::bcx(UGeckoInstruction) + 652
    frame #2: 0x0000000100c3070c Dolphin`JitArm64::CompileInstruction(PPCAnalyst::CodeOp&) + 104
    frame #3: 0x0000000100bdf298 Dolphin`JitArm64::DoJit(unsigned int, JitBlock*, unsigned int) + 2964
    frame #4: 0x0000000100bddc80 Dolphin`JitArm64::Jit(unsigned int, bool) + 848
    frame #5: 0x0000000100bdca88 Dolphin`JitArm64::Jit(unsigned int) + 40
    frame #6: 0x0000000100b1f9dc Dolphin`JitTrampoline(JitBase&, unsigned int) + 40
    frame #7: 0x00000002953e50c0
```

JosJuice commented:
> The dtm file doesn’t depend on a save state so rebasing is fine.

My concern wasn't about savestates not loading, it was about the inputs desyncing. But okay, if it syncs for you after rebasing, I suppose we haven't had any sync-breaking changes since you made the DTM.

I've rebased the PR and added a patch that should fix the assert you reported getting. Does it also fix the freeze, or is that still happening? If it's still happening, could you bisect within the PR?

dvessel commented:
It works! Thanks for the fix.

krnlyng commented on Source/Core/Core/PowerPC/JitArm64/Jit.cpp:
```diff
@@ -363,18 +363,35 @@ void JitArm64::WriteExit(u32 destination, bool LK, u32 exit_address_after_return
+    // Push {ARM_PC; PPC_PC} on the stack
     MOVI2R(ARM64Reg::X1, exit_address_after_return);
-    ADR(ARM64Reg::X0, 20);
+    ADR(ARM64Reg::X0, JitArm64BlockCache::BLOCK_LINK_FAST_BL_OFFSET + 12);
```
The 12 is a bit of a magic value, it's used in a few places, does it warrant a macro/constexpr or some ASSERT that the distance to after the BL is still 12, an it could maybe also be put in terms of 3 * sizeof-arm64-instruction?

JosJuice replied:
I've rewritten this one as `JitArm64BlockCache::BLOCK_LINK_SIZE + sizeof(u32) * 2`, which should hopefully make it clearer. The 2 is for the 2 instructions here before the block link: The ADR and the STP. I've added asserts too.

krnlyng commented on Source/Core/Core/PowerPC/JitArm64/Jit.cpp:
```diff
@@ -838,18 +915,7 @@ bool JitArm64::DoJit(u32 em_address, JitBlock* b, u32 nextPC)
   js.numFloatingPointInst = 0;
 
   u8* const start = GetWritableCodePtr();
```
Is there still much use for the temporary start variable? You could just set normalEntry directly and use normalEntry in the codeSize computation later, like in CachedInterpreter.cpp.

JosJuice replied:
Done.

krnlyng commented on Source/Core/Core/PowerPC/JitArm64/Jit.cpp:
```diff
@@ -363,18 +363,35 @@ void JitArm64::WriteExit(u32 destination, bool LK, u32 exit_address_after_return
+  const u8* farcode_1_addr;
+  if (switch_to_far_code)
+  {
+    SwitchToFarCode();
```
Wouldn't farcode_1_addr = m_far_code.GetCodePtr() be simpler? Also not sure if there are more descriptive names than farcode_1_addr and farcode_2_addr?

JosJuice replied:
I didn't think of `m_far_code.GetCodePtr()`. But unfortunately, after testing it, it doesn't actually work for what we want to do. Calling `SwitchToFarCode` will align the current farcode pointer to 16 bytes, and we need to take that into account when obtaining the address.

I've renamed the variables.

krnlyng commented on Source/Core/Core/PowerPC/JitArm64/Jit.cpp:
```diff
@@ -363,18 +363,35 @@ void JitArm64::WriteExit(u32 destination, bool LK, u32 exit_address_after_return
     STP(IndexType::Pre, ARM64Reg::X0, ARM64Reg::X1, ARM64Reg::SP, -16);
   }
 
+  constexpr size_t farcode_size = 3 * sizeof(u32);
```
Isn't this 3 * sizeof(u32) also the same as JitArm64BlockCache::BLOCK_LINK_SIZE?

JosJuice replied:
Yes, but that is a coincidence.

krnlyng commented on Source/Core/Core/PowerPC/JitArm64/JitArm64Cache.cpp:
```diff
@@ -22,64 +22,81 @@ void JitArm64BlockCache::Init()
   {
-    emit.B(CC_GT, dest->normalEntry);
-    emit.B(dest->checkedEntry);
     emit.BRK(101);
```
Should there be an ASSERT(!source.call) in the loop? Because if we RET() to a BRK things would go wrong.

krnlyng commented on Source/Core/Core/PowerPC/JitArm64/Jit.cpp:
```diff
@@ -386,10 +403,32 @@ void JitArm64::WriteExit(u32 destination, bool LK, u32 exit_address_after_return
+  if (LK)
+  {
+    if (GetCodePtr() == farcode_2_addr - sizeof(u32))
+      BRK(101);
```
Wouldn't we end up executing that BRK if we return from the BL above?

Also i've noted that in some calls to BRK you're using a hex value and in some a decimal. But i don't think it matters.

JosJuice replied:
> Ah the BRK is obviously to fill the 3 instructions so that the return value pushed to the stack is correct. NVM!

Actually, that's not it. Before this PR, `JitArm64BlockCache::WriteLinkBlock` would emit two BLs: The "slow" BL and the "fast" BL. The fast BL matched the address pushed to the stack, but the slow one didn't, so if the slow BL were to be triggered, the BLR optimization would fail. What has happened in this commit is that the slow BL was moved to the farcode you see here. Just like before, it does not match the address pushed to the stack. And because of that, we'll never actually return to here.

The actual reason why we need this is because we made a "prediction" earlier that `farcode_size` would be 3 instructions long. If it turned out to be 2 instructions long (`MOVI2R` can emit either 1 or 2 instructions in this situation), we just pad it so it's 3 instructions long. It wouldn't be impossible to make `farcode_size` correctly predict whether we get 2 or 3 instructions, but it seems unnecessarily complicated to me.

krnlyng reviewed (commented):
I've added some minor comments. Feel free to completely ignore me though.
